src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs

```diff
@@ -156,87 +56,58 @@ public override int Order
             {
                 throw new ArgumentNullException(nameof(output));
             }
+            
+            IHtmlContent content = null;
 
-            IHtmlContent result = null;
             if (Enabled)
             {
                 var key = GenerateKey(context);
-                if (!MemoryCache.TryGetValue(key, out result))
+                MemoryCacheEntryOptions options;
+    
+                while(content == null)
                 {
-                    // Create an entry link scope and flow it so that any tokens related to the cache entries
-                    // created within this scope get copied to this scope.
-                    using (var link = MemoryCache.CreateLinkingScope())
+                    Task<IHtmlContent> result = null;
+                    
+                    if (!MemoryCache.TryGetValue(key, out result))
                     {
-                        var content = await output.GetChildContentAsync();
-
-                        var stringBuilder = new StringBuilder();
-                        using (var writer = new StringWriter(stringBuilder))
+                        // Create an entry link scope and flow it so that any tokens related to the cache entries
+                        // created within this scope get copied to this scope.
+                        options = GetMemoryCacheEntryOptions();
+
+                        var tcs = new TaskCompletionSource<IHtmlContent>();
+                         
+                        MemoryCache.Set(key, tcs.Task, options);
+                        
+                        using (var link = MemoryCache.CreateLinkingScope())
                         {
-                            content.WriteTo(writer, HtmlEncoder);
+                            result = ProcessContentAsync(output);
+                            content = await result;
+                            options.AddEntryLink(link);
                         }
-
-                        result = new StringBuilderHtmlContent(stringBuilder);
-                        MemoryCache.Set(key, result, GetMemoryCacheEntryOptions(link));
+                        
+                        MemoryCache.Set(key, result, options);
```

Comment:
I am adding a code comment for clarification. The idea is that the second setter will use the actual options, so that the expiration is not impacted by the time it takes to render the content.